File: src/plugins/countdown.ts
import { IMatrixClient, Plugin, CommandDef, MessageContext } from "./base";
import { getDb } from "../db";
import logger from "../utils/logger";

const MONTH_NAMES = [
  "Jan", "Feb", "Mar", "Apr", "May", "Jun",
  "Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
];

function formatDate(dateStr: string): string {
  const d = new Date(dateStr + "T00:00:00Z");
  const month = MONTH_NAMES[d.getUTCMonth()];
  const day = d.getUTCDate();
  const year = d.getUTCFullYear();
  return `${month} ${day}, ${year}`;
}

function todayStr(): string {
  const now = new Date();
  const y = now.getUTCFullYear();
  const m = String(now.getUTCMonth() + 1).padStart(2, "0");
  const d = String(now.getUTCDate()).padStart(2, "0");
  return `${y}-${m}-${d}`;
}

function daysBetween(a: string, b: string): number {
  const msA = new Date(a + "T00:00:00Z").getTime();
  const msB = new Date(b + "T00:00:00Z").getTime();
  return Math.round((msB - msA) / (1000 * 60 * 60 * 24));
}

interface CountdownRow {
  id: number;
  user_id: string;
  room_id: string;
  label: string;
  target_date: string;
  public: number;
  completed: number;
  created_at: number;
}

export class CountdownPlugin extends Plugin {
  constructor(client: IMatrixClient) {
    super(client);
  }

  get name() {
    return "countdown";
  }

  get commands(): CommandDef[] {
    return [
      { name: "countdown add", description: "Add a public countdown", usage: '!countdown add "<label>" <YYYY-MM-DD>' },
      { name: "countdown private", description: "Add a private countdown", usage: '!countdown private "<label>" <YYYY-MM-DD>' },
      { name: "countdown mine", description: "List your own countdowns" },
      { name: "countdown remove", description: "Remove a countdown", usage: "!countdown remove <id>" },
      { name: "countdown", description: "List countdowns or show a specific one", usage: "!countdown [id]" },
    ];
  }

  async onMessage(ctx: MessageContext): Promise<void> {
    if (this.isCommand(ctx.body, "countdown add")) {
      await this.handleAdd(ctx, true);
    } else if (this.isCommand(ctx.body, "countdown private")) {
      await this.handleAdd(ctx, false);
    } else if (this.isCommand(ctx.body, "countdown mine")) {
      await this.handleMine(ctx);
    } else if (this.isCommand(ctx.body, "countdown remove")) {
      await this.handleRemove(ctx);
    } else if (this.isCommand(ctx.body, "countdown")) {
      const args = this.getArgs(ctx.body, "countdown").trim();
      if (args && /^\d+$/.test(args)) {
        await this.handleShow(ctx, parseInt(args));
      } else if (!args) {
        await this.handleList(ctx);
      }
    }
  }

  private parseAddArgs(raw: string): { label: string; date: string } | null {
    // Expect: "<label>" <YYYY-MM-DD>
    const match = raw.match(/^"([^"]+)"\s+(\d{4}-\d{2}-\d{2})$/);
    if (!match) return null;
    return { label: match[1], date: match[2] };
  }

  private isValidDate(dateStr: string): boolean {
    const d = new Date(dateStr + "T00:00:00Z");
    return !isNaN(d.getTime());
  }

  private async handleAdd(ctx: MessageContext, isPublic: boolean): Promise<void> {
    const command = isPublic ? "countdown add" : "countdown private";
    const args = this.getArgs(ctx.body, command).trim();
    const parsed = this.parseAddArgs(args);

    if (!parsed) {
      const usage = isPublic
        ? '!countdown add "<label>" <YYYY-MM-DD>'
        : '!countdown private "<label>" <YYYY-MM-DD>';
      await this.sendReply(ctx.roomId, ctx.eventId, `Usage: ${usage}`);
      return;
    }

    if (!this.isValidDate(parsed.date)) {
      await this.sendReply(ctx.roomId, ctx.eventId, "Invalid date format. Use YYYY-MM-DD.");
      return;
    }

    const db = getDb();
    const result = db.prepare(
      `INSERT INTO countdowns (user_id, room_id, label, target_date, public) VALUES (?, ?, ?, ?, ?)`
    ).run(ctx.sender, ctx.roomId, parsed.label, parsed.date, isPublic ? 1 : 0);

    const id = result.lastInsertRowid;
    const today = todayStr();
    const days = daysBetween(today, parsed.date);
    const direction = days >= 0 ? `${days} days away` : `${Math.abs(days)} days ago`;
    const visibility = isPublic ? "public" : "private";

    await this.sendReply(
      ctx.roomId,
      ctx.eventId,
      `Countdown #${id} added (${visibility}): "${parsed.label}" on ${formatDate(parsed.date)} — ${direction}.`
    );
  }

  private async handleRemove(ctx: MessageContext): Promise<void> {
    const idStr = this.getArgs(ctx.body, "countdown remove").trim();
    if (!idStr || !/^\d+$/.test(idStr)) {
      await this.sendReply(ctx.roomId, ctx.eventId, "Usage: !countdown remove <id>");
      return;
    }

    const id = parseInt(idStr);
    const db = getDb();
    const result = db.prepare(
      `DELETE FROM countdowns WHERE id = ? AND user_id = ?`
    ).run(id, ctx.sender);

    if (result.changes > 0) {
      await this.sendReply(ctx.roomId, ctx.eventId, `Countdown #${id} removed.`);
    } else {
      await this.sendReply(ctx.roomId, ctx.eventId, `No countdown found with ID ${id} (or it's not yours).`);
    }
  }

  private async handleShow(ctx: MessageContext, id: number): Promise<void> {
    const db = getDb();
    const row = db.prepare(
      `SELECT * FROM countdowns WHERE id = ? AND room_id = ? AND completed = 0`
    ).get(id, ctx.roomId) as CountdownRow | undefined;

    if (!row) {
      await this.sendReply(ctx.roomId, ctx.eventId, `No active countdown found with ID ${id}.`);
      return;
    }

    // Only show private countdowns to their owner
    if (!row.public && row.user_id !== ctx.sender) {
      await this.sendReply(ctx.roomId, ctx.eventId, `No active countdown found with ID ${id}.`);
      return;
    }

    const today = todayStr();
    const days = daysBetween(today, row.target_date);
    const visibility = row.public ? "public" : "private";

    let detail: string;
    if (days > 0) {
      detail = `#${row.id} "${row.label}" — ${days} days  (${formatDate(row.target_date)}) [${visibility}]`;
    } else if (days === 0) {
      detail = `#${row.id} "${row.label}" — today!  (${formatDate(row.target_date)}) [${visibility}]`;
    } else {
      detail = `#${row.id} "${row.label}" — happened ${Math.abs(days)} days ago  (${formatDate(row.target_date)}) [${visibility}]`;
    }

    await this.sendMessage(ctx.roomId, detail);
  }

  private async handleList(ctx: MessageContext): Promise<void> {
    const db = getDb();
    const today = todayStr();

    // Fetch all public countdowns for this room + sender's private ones, not yet completed
    const rows = db.prepare(
      `SELECT * FROM countdowns
       WHERE room_id = ? AND completed = 0
         AND (public = 1 OR user_id = ?)
       ORDER BY target_date ASC`
    ).all(ctx.roomId, ctx.sender) as CountdownRow[];

    // Cleanup: mark completed any that are past > 7 days
    const toComplete: number[] = [];
    const visible: CountdownRow[] = [];

    for (const row of rows) {
      const days = daysBetween(today, row.target_date);
      if (days < 0 && Math.abs(days) > 7) {
        toComplete.push(row.id);
      } else {
        visible.push(row);
      }
    }

    if (toComplete.length > 0) {
      const placeholders = toComplete.map(() => "?").join(",");
      db.prepare(`UPDATE countdowns SET completed = 1 WHERE id IN (${placeholders})`).run(...toComplete);
    }

    if (visible.length === 0) {
      await this.sendReply(ctx.roomId, ctx.eventId, "No active countdowns.");
      return;
    }

    const lines = this.formatCountdownLines(visible, today);
    await this.sendMessage(ctx.roomId, `Community Countdowns\n\n${lines.join("\n")}`);
  }

  private async handleMine(ctx: MessageContext): Promise<void> {
    const db = getDb();
    const today = todayStr();

    const rows = db.prepare(
      `SELECT * FROM countdowns
       WHERE user_id = ? AND room_id = ? AND completed = 0
       ORDER BY target_date ASC`
    ).all(ctx.sender, ctx.roomId) as CountdownRow[];

    // Cleanup: mark completed any that are past > 7 days
    const toComplete: number[] = [];
    const visible: CountdownRow[] = [];

    for (const row of rows) {
      const days = daysBetween(today, row.target_date);
      if (days < 0 && Math.abs(days) > 7) {
        toComplete.push(row.id);
      } else {
        visible.push(row);
      }
    }

    if (toComplete.length > 0) {
      const placeholders = toComplete.map(() => "?").join(",");
      db.prepare(`UPDATE countdowns SET completed = 1 WHERE id IN (${placeholders})`).run(...toComplete);
    }

    if (visible.length === 0) {
      await this.sendReply(ctx.roomId, ctx.eventId, "You have no active countdowns.");
      return;
    }

    const lines = this.formatCountdownLines(visible, today);
    await this.sendMessage(ctx.roomId, `Your Countdowns\n\n${lines.join("\n")}`);
  }

  private formatCountdownLines(rows: CountdownRow[], today: string): string[] {
    return rows.map((row) => {
      const days = daysBetween(today, row.target_date);
      const label = row.label;

      if (days > 0) {
        const pad = " ".repeat(Math.max(0, 30 - label.length));
        return `${label}${pad} — ${days} days  (${formatDate(row.target_date)})`;
      } else if (days === 0) {
        const pad = " ".repeat(Math.max(0, 30 - label.length));
        return `${label}${pad} — today!  (${formatDate(row.target_date)})`;
      } else {
        const ago = Math.abs(days);
        const padLabel = `\u2705 ${label}`;
        const pad = " ".repeat(Math.max(0, 30 - padLabel.length));
        return `${padLabel}${pad} — happened ${ago} days ago`;
      }
    });
  }
}

File: internal/plugin/countdown.go
package plugin

import (
	"fmt"
	"log/slog"
	"regexp"
	"strconv"
	"strings"
	"time"

	"gogobee/internal/db"

	"maunium.net/go/mautrix"
)

var countdownLabelRe = regexp.MustCompile(`"([^"]+)"\s+(\d{4}-\d{2}-\d{2})`)

// CountdownPlugin manages countdown timers to specific dates.
type CountdownPlugin struct {
	Base
}

// NewCountdownPlugin creates a new countdown plugin.
func NewCountdownPlugin(client *mautrix.Client) *CountdownPlugin {
	return &CountdownPlugin{
		Base: NewBase(client),
	}
}

func (p *CountdownPlugin) Name() string { return "countdown" }

func (p *CountdownPlugin) Commands() []CommandDef {
	return []CommandDef{
		{Name: "countdown", Description: "Manage countdowns to dates", Usage: "!countdown [add/private/mine/remove/id]", Category: "Personal"},
	}
}

func (p *CountdownPlugin) Init() error {
	// Auto-complete old countdowns
	p.completeOldCountdowns()
	return nil
}

func (p *CountdownPlugin) OnReaction(_ ReactionContext) error { return nil }

func (p *CountdownPlugin) OnMessage(ctx MessageContext) error {
	if !p.IsCommand(ctx.Body, "countdown") {
		return nil
	}

	args := strings.TrimSpace(p.GetArgs(ctx.Body, "countdown"))

	switch {
	case strings.HasPrefix(strings.ToLower(args), "add "):
		return p.handleAdd(ctx, args[4:], true)
	case strings.HasPrefix(strings.ToLower(args), "private "):
		return p.handleAdd(ctx, args[8:], false)
	case strings.ToLower(args) == "mine":
		return p.handleMine(ctx)
	case strings.HasPrefix(strings.ToLower(args), "remove "):
		return p.handleRemove(ctx, strings.TrimSpace(args[7:]))
	case args == "":
		return p.handleList(ctx)
	default:
		// Try to parse as an ID
		if _, err := strconv.Atoi(args); err == nil {
			return p.handleShow(ctx, args)
		}
		return p.SendReply(ctx.RoomID, ctx.EventID,
			"Usage: !countdown add \"<label>\" <YYYY-MM-DD> | private \"<label>\" <YYYY-MM-DD> | mine | remove <id> | [id]")
	}
}

func (p *CountdownPlugin) handleAdd(ctx MessageContext, input string, public bool) error {
	matches := countdownLabelRe.FindStringSubmatch(input)
	if matches == nil {
		return p.SendReply(ctx.RoomID, ctx.EventID,
			"Usage: !countdown add \"<label>\" <YYYY-MM-DD>")
	}

	label := matches[1]
	dateStr := matches[2]

	_, err := time.Parse("2006-01-02", dateStr)
	if err != nil {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Invalid date format. Use YYYY-MM-DD.")
	}

	d := db.Get()
	isPublic := 1
	if !public {
		isPublic = 0
	}

	_, err = d.Exec(
		`INSERT INTO countdowns (user_id, room_id, label, target_date, public) VALUES (?, ?, ?, ?, ?)`,
		string(ctx.Sender), string(ctx.RoomID), label, dateStr, isPublic,
	)
	if err != nil {
		slog.Error("countdown: add", "err", err)
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to add countdown.")
	}

	visibility := "public"
	if !public {
		visibility = "private"
	}
	return p.SendReply(ctx.RoomID, ctx.EventID,
		fmt.Sprintf("Countdown added (%s): \"%s\" on %s", visibility, label, dateStr))
}

func (p *CountdownPlugin) handleMine(ctx MessageContext) error {
	d := db.Get()
	rows, err := d.Query(
		`SELECT id, label, target_date, public FROM countdowns
		 WHERE user_id = ? AND completed = 0
		 ORDER BY target_date ASC`,
		string(ctx.Sender),
	)
	if err != nil {
		slog.Error("countdown: mine", "err", err)
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to load your countdowns.")
	}
	defer rows.Close()

	var sb strings.Builder
	sb.WriteString("Your countdowns:\n\n")
	count := 0
	now := time.Now().UTC()

	for rows.Next() {
		var cdID int
		var label, targetDate string
		var public int
		if err := rows.Scan(&cdID, &label, &targetDate, &public); err != nil {
			continue
		}
		t, _ := time.Parse("2006-01-02", targetDate)
		days := int(t.Sub(now).Hours() / 24)
		vis := ""
		if public == 0 {
			vis = " (private)"
		}
		status := fmt.Sprintf("%d days", days)
		if days < 0 {
			status = fmt.Sprintf("%d days ago", -days)
		} else if days == 0 {
			status = "TODAY!"
		}
		sb.WriteString(fmt.Sprintf("  [%d] %s — %s (%s)%s\n", cdID, label, targetDate, status, vis))
		count++
	}

	if count == 0 {
		return p.SendReply(ctx.RoomID, ctx.EventID, "You have no active countdowns.")
	}

	return p.SendReply(ctx.RoomID, ctx.EventID, sb.String())
}

func (p *CountdownPlugin) handleRemove(ctx MessageContext, idStr string) error {
	d := db.Get()
	result, err := d.Exec(
		`DELETE FROM countdowns WHERE id = ? AND user_id = ?`,
		idStr, string(ctx.Sender),
	)
	if err != nil {
		slog.Error("countdown: remove", "err", err)
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to remove countdown.")
	}

	affected, _ := result.RowsAffected()
	if affected == 0 {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Countdown not found or not yours.")
	}

	return p.SendReply(ctx.RoomID, ctx.EventID, "Countdown removed.")
}

func (p *CountdownPlugin) handleList(ctx MessageContext) error {
	d := db.Get()
	rows, err := d.Query(
		`SELECT id, label, target_date, user_id FROM countdowns
		 WHERE public = 1 AND completed = 0
		 ORDER BY target_date ASC LIMIT 20`,
	)
	if err != nil {
		slog.Error("countdown: list", "err", err)
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to load countdowns.")
	}
	defer rows.Close()

	var sb strings.Builder
	sb.WriteString("Active countdowns:\n\n")
	count := 0
	now := time.Now().UTC()

	for rows.Next() {
		var cdID int
		var label, targetDate, userID string
		if err := rows.Scan(&cdID, &label, &targetDate, &userID); err != nil {
			continue
		}
		t, _ := time.Parse("2006-01-02", targetDate)
		days := int(t.Sub(now).Hours() / 24)
		status := fmt.Sprintf("%d days", days)
		if days < 0 {
			status = fmt.Sprintf("%d days ago", -days)
		} else if days == 0 {
			status = "TODAY!"
		}
		sb.WriteString(fmt.Sprintf("  [%d] %s — %s (%s) by %s\n", cdID, label, targetDate, status, userID))
		count++
	}

	if count == 0 {
		return p.SendReply(ctx.RoomID, ctx.EventID, "No active countdowns. Add one with !countdown add \"<label>\" <date>")
	}

	return p.SendReply(ctx.RoomID, ctx.EventID, sb.String())
}

func (p *CountdownPlugin) handleShow(ctx MessageContext, idStr string) error {
	d := db.Get()
	var label, targetDate, userID string
	var public int
	err := d.QueryRow(
		`SELECT label, target_date, user_id, public FROM countdowns WHERE id = ? AND completed = 0`,
		idStr,
	).Scan(&label, &targetDate, &userID, &public)
	if err != nil {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Countdown not found.")
	}

	// Private countdowns only visible to owner
	if public == 0 && userID != string(ctx.Sender) {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Countdown not found.")
	}

	t, _ := time.Parse("2006-01-02", targetDate)
	now := time.Now().UTC()
	diff := t.Sub(now)

	var status string
	if diff < 0 {
		absDays := int(-diff.Hours()) / 24
		status = fmt.Sprintf("%d days ago", absDays)
	} else {
		days := int(diff.Hours()) / 24
		hours := int(diff.Hours()) % 24
		if days == 0 {
			status = fmt.Sprintf("%d hours remaining!", hours)
		} else {
			status = fmt.Sprintf("%d days and %d hours", days, hours)
		}
	}

	msg := fmt.Sprintf("\"%s\" — %s\n%s (by %s)", label, targetDate, status, userID)
	return p.SendReply(ctx.RoomID, ctx.EventID, msg)
}

func (p *CountdownPlugin) completeOldCountdowns() {
	d := db.Get()
	cutoff := time.Now().UTC().AddDate(0, 0, -7).Format("2006-01-02")
	_, err := d.Exec(
		`UPDATE countdowns SET completed = 1 WHERE target_date < ? AND completed = 0`,
		cutoff,
	)
	if err != nil {
		slog.Error("countdown: auto-complete", "err", err)
	}
}
